artifact/scripts/executor.cpp

```diff
@@ -0,0 +1,252 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <artifact/scripts/executor.hpp>
+
+#include <algorithm>
+#include <chrono>
+#include <regex>
+#include <string>
+
+#include <common/common.hpp>
+#include <common/expected.hpp>
+#include <common/path.hpp>
+
+
+namespace mender {
+namespace artifact {
+namespace scripts {
+namespace executor {
+
+namespace expected = mender::common::expected;
+
+
+using expected::ExpectedBool;
+
+namespace common = mender::common;
+namespace merror = mender::common::error;
+namespace path = mender::common::path;
+
+
+const int expected_state_script_version {3};
+
+unordered_map<const State, string> state_map {
+	{State::Idle, "Idle"},
+	{State::Sync, "Sync"},
+	{State::Download, "Download"},
+	{State::ArtifactInstall, "ArtifactInstall"},
+	{State::ArtifactReboot, "ArtifactReboot"},
+	{State::ArtifactCommit, "ArtifactCommit"},
+	{State::ArtifactRollback, "ArtifactRollback"},
+	{State::ArtifactRollbackReboot, "ArtifactRollbackReboot"},
+	{State::ArtifactFailure, "ArtifactFailure"},
+};
+
+unordered_map<const Action, string> action_map {
+	{Action::Enter, "Enter"},
+	{Action::Leave, "Leave"},
+	{Action::Error, "Error"},
+};
+
+merror::Error CorrectVersionFile(const string &path) {
+	ifstream vf {path};
+
+	// Missing file is OK
```

Comment:
This should be explained, I think. It's because Golang did not have a version file, so no-file is equal to v3.